### ISSUE TITLE
What is the lsp-snippet story?

### ISSUE THREAD

**Issue by Avi-D-coder** [ul/kak-lsp#145]
Is there a setup that enables lsp-snippet completions?
https://discuss.kakoune.com/t/need-snippets-we-got-snippets/42 mentions several snippet plugins, but no confirmation of a working setup.

Does kak-lsp emit snippets?
If not what is blocking this use case?

**Comment by ul** [comment]
I think as soon as we add `snippetSupport: true` into client capabilities of initialize request kak-lsp will automagically be able to emit snippets provided by the language server.

Then there is a problem of handling those snippets in Kakoune. https://github.com/JJK96/kakoune-snippets seems to be the closest in snippet format, but it could require some extra setup to trigger on completion.

If kak-lsp will rely on an external plugin to handle snippets then `snippetSupport` client capability should be configurable.

**Comment by JJK96** [comment]
I have a local branch of `kak-lsp` in which snippets are enabled. 

The main problem is that currently, my plugin searched forward from the cursor position to the next hole to be filled. So, if autocompletion is used to insert a snippet the cursor is behind it and has to search through the whole file for the hole (thereby possible encountering other strings that match the hole syntax). 

I think that a good solution for this would be to have a hook executed on `InsertCompletionHide` which would then select the inserted text and check if a snippet hole exists inside the selection and if so try to fill it. I haven't yet found a good way to select the inserted completion.

Probably [This issue](https://github.com/mawww/kakoune/issues/2347) is relevant.

**Comment by andreyorst** [comment]
> Then there is a problem of handling those snippets in Kakoune. https://github.com/JJK96/kakoune-snippets seems to be the closest in snippet format, but it could require some extra setup to trigger on completion.

@occivink started really promicing snippet plugin: [kakoune-snippets](https://github.com/occivink/kakoune-snippets). It already supports standard snippet syntax (partly).

**Comment by JJK96** [comment]
Yeah, I have seen it, I will probably start supporting his plugin instead of mine, however I expect that his plugin will run into the same problem that I mentioned above.

**Comment by occivink** [comment]
I don't think this is going to be a problem, currently the plugin works by pasting with `<a-P>` at the cursor location and then finding the placeholders inside with `s`.

**Comment by JJK96** [comment]
The problem is that snippets from kak-lsp are entered through yhe autocomplete menu, thus, they are not pasted, so you cannot easily search within that range.

**Comment by ul** [comment]
Also if we want to pass snippet directly to the snippet manager then this might be helpful https://github.com/mawww/kakoune/issues/2347

**Comment by JJK96** [comment]
Inserted snippets without a way to handle them are just annoying, so the user should be able to enable/disable. I think it would be a good idea to also provide an option which takes a command to handle the snippet so that kak-lsp does not need to depend on a particular snippet manager.

**Comment by ul** [comment]
Now with completions-commands branch of Kakoune we can make a proper integration with snippet managers. E.g. we can introduce option like `lsp_snippet_callback` which will be evaluated with snippet content as a parameter instead of inserting it straight away.

**Comment by andreyorst** [comment]
that's neat. I guess it's OK for kakoune to use completion system for snippets, since it is builtin, where in Vim, builtin completion isn't capable to di such ootb

**Comment by ul** [comment]
I did the first step and added an option to control snippet support client capability. So if you have some working integration for snippets inserted as the result of completion, you can set `snippet_support = true` in `kak-lsp.toml` to ask language servers to provide snippets where applicable. The next step would be to design and implement something like `lsp_snippet_callback` to allow snippet plugins to not guess when to kick in but be called explicitly.

**Comment by JJK96** [comment]
I've been trying a bit with completions and snippets, I have the following proof of concept:

Buffer:
```

declare-option completions test
hook window InsertIdle .* %{
    set-option window test "1.10@%val{timestamp}" 'a|exec -draft u;snippets-insert %{snippet a ($1)}|a' 'b|exec -draft u;snippets-insert %{snippet b ($1)}|b'
}
hook window InsertCompletionShow .* %{
    execute-keys -draft i<c-u>
}
set-option window completers option=test
```

Steps:
- install [kakoune-snippets](https://github.com/occivink/kakoune-snippets)
- source %val{buffile}
- Start typing on the first line until the completion menu is shown.
- Try selecting completions

Currently one problem, after cycling through all completions, there is still some leftover text. I'm not sure how to resolve that, maybe the cleanup command can help. Anyway, pressing <kbd>u</kbd> removes it for you

**Comment by andreyorst** [comment]
Shouldn't snippets be inserted after confirming completion?

**Comment by JJK96** [comment]
I think inserting them immediately is nicer

**Comment by andreyorst** [comment]
if user wants to use the same key for completing and jumping through snippets this may cause an issue.

Can you show in ascciicast or gif of how snippets are inserted in your code when cycling completions?

**Comment by JJK96** [comment]
Asciicast: https://asciinema.org/a/DqmFDowao0kQMTcNUlixyibuY
The colors are broken, I don't know why

**Comment by ul** [comment]
I think this is mostly resolved now by #270. Feel free to open new issues for improvements over the current solution.
